### ISSUE TITLE
Strange behavior in `nx.effective_size` for the single node graph

### ISSUE THREAD

**Issue by joyemang33** [networkx/networkx#6916]
Hello! It seems I've created quite a few issue reports, and I apologize for any inconvenience. Please feel free to review them at your own pace, or you can assign them to me if I can assist in creating PR to fix them.

According to the formula
![image](https://github.com/networkx/networkx/assets/37923722/ca45adbf-b793-41cc-83b4-a8bd01be22ce)
The `effective_size` should have a well-defined value of `0` for a graph with a single node.
However, if we use Borgatti's method to calculate this graph, the value will lose its definition, and NetworkX will return some strange behavior:

For instance:
```Python
import networkx as nx

G = nx.Graph()
G.add_node(0)
print(nx.effective_size(G))

G = nx.Graph([(0, 0)])
print(nx.effective_size(G))
```
#### Result:
```
{0: nan}
ZeroDivisionError
```
I am not sure whether such behaviors are expected. So it would be highly appreciated if you could further confirm and investigate it.
If they are unexpected, I will create a PR to fix it by returning a single `0` if the graph contains only one node.

Best regards,
Joye

### Environment
NetworkX 3.1
Python 3.10

**Comment by MridulS** [comment]
>  If they are unexpected, I will create a PR to fix it by returning a single 0 if the graph contains only one node.

Yeah we should be catching this corner case. Thanks for your work on this!! Please do send a PR :)

**Comment by MridulS** [comment]
Looking at this one again a code comment explicitly states that for the corner case of isolated nodes while using Borgatti's formula it should be undefined (`nan`) 
https://github.com/networkx/networkx/blob/f35e034d33864201e30d53edc37f55f38b384d94/networkx/algorithms/structuralholes.py#L147-L149

But the other case is where I guess we need to make a choice. I think for consistency we put in undefined behavior for isolated nodes but according to the reference it should be zero.
https://github.com/networkx/networkx/blob/f35e034d33864201e30d53edc37f55f38b384d94/networkx/algorithms/structuralholes.py#L155-L157

Also the ZeroDivisionError is popping up because of the self loop which makes the node `0` **not** isolated and hence the zero division error isn't caught early on, well I guess the question is does a node with a self loop and no other connection an isolated node?

**Comment by joyemang33** [comment]
Hi @MridulS. Thanks for your kind help in this issue. 
In my opinion, a node with a self-loop and no other connection may need to be regarded as an isolated node, and we need to give a zero value in this case. This is because its `effective_size` computed by the above equation will still be zero (N(u) \ u = empty set).

That's only my personal opinion, so please feel free to tell me if I misunderstand the question.

Best regards,
Joye

**Comment by rossbar** [comment]
IIUC, it seems that the returning `nan` for isolated nodes is agreed upon as the correct, or at least sensible, behavior. So the remaining question is isolated nodes with self-loops. I'm no expert, but looking at Borgatti's formula in the docstring, it seems that an isolated node with a self-edge would still result in `0 - 0/0`, and therefore should also result in `nan`. Absent any explicit information on how self-loops are treated from the Borgatti/Burt papers, this is what I think makes the most sense.
